src/Microsoft.AspNet.Http.Abstractions/Extensions/MapWhenOptions.cs

```diff
@@ -13,14 +11,26 @@ namespace Microsoft.AspNet.Builder.Extensions
     /// </summary>
     public class MapWhenOptions
     {
+        private Func<HttpContext, bool> _predicate;
+
         /// <summary>
         /// The user callback that determines if the branch should be taken
         /// </summary>
         public Func<HttpContext, bool> Predicate
         {
-            get;
-            [param: NotNull]
-            set;
+            get
+            {
+                return _predicate;
+            }
+            set
+            {
+                if (value == null)
```

Comment:
Keep the Options simple and move this validation into the middleware constructor.